spring-kafka/src/main/java/org/springframework/kafka/listener/ContainerFencedException.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2018-2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Exception throw when container is started when it is not a member of
+ * ConcurrentMessageListenerContainer.
```

Comment:
I'm not sure in this conclusion.
The `KafkaMessageListenerContainer` can be used by itself, without being a member of the `ConcurrentMessageListenerContainer`.

I think we talked about failing to start a child container when its parent is stopped.
Not sure also if we definitely need a dedicated exception on the matter.
I feel like regular `IllegalStateException` is fully enough.
Because indeed we are in the illegal state to attempt to start that child container when its parent is stopped.

What do I miss, please?

Reply:
From design perspective I thought, It would be better if we can provide a specific exception for this scenario. But, `IllegalStateException` would also works here. I will do that change here if you suggest.

Reply:
I think new exception would be benefical for the app developers. It improves the error handling. IllegalStateException would work, but it is really generic. From framework perspective, a seperate exception could be better. Please give your comments. I will modify accordingly.

Reply:
> The `KafkaMessageListenerContainer` can be used by itself, without being a member of the `ConcurrentMessageListenerContainer`.
> 
> I think we talked about failing to start a child container when its parent is stopped.

KafkaMessageListenerContainer can be used individually. Given changes does not change any modify in this regard. I feel expected spring-kafka system should be as follows.

Assumption: CMain -- ConcurrentContainer. C0, C1 are the child containers. Choosen concurrency 2.

Flow:

1) CMain -- started. 
   This would inturn starts C0 and C1.
   
   CMain -- running.
   C0 -- running
   C1 -- running.
   
2) C0 stopped. This should be allowed.
   
   CMain -- not running.
   C0 -- not running
   C1 -- running.
   
3) C0 started. This should be allowed. Since, CMain is not stopped explicitly.
   
   CMain -- running.
   C0 -- running
   C1 -- running.

4) C0 stopped. This should be allowed. Since, CMain is not stopped explicitly.
   
   CMain -- not running.
   C0 -- not running
   C1 -- running.

5) C1 stopped. This should be allowed. Since, CMain is not stopped explicitly.
   
   CMain -- not running.
   C0 -- not running
   C1 -- not running.

6) C0 started. This should be allowed. Since, CMain is not stopped explicitly.
   
   CMain -- not running.
   C0 -- running
   C1 -- not running.

7) C1 started. This should be allowed. Since, CMain is not stopped explicitly.
   
   CMain -- running.
   C0 -- running
   C1 -- running.

8) CMain -- stopped explicitly. 
   This would in turn stops C0 and C1. C0 and C1 would be cleared.
   
   CMain -- not running.
   C0 -- not running. delinked
   C1 -- not running. delinked.

9) **C0 started. This should not be allowed. Since, CMain is stopped explicitly. Container Should be Fenced.**
   

10) CMain -- started. 
   This would inturn starts C2 and C3.
   
   CMain -- running.
   C2 -- running
   C3 -- running.

   C0 -- not running. delinked
   C1 -- not running. delinked.

Reply:
The custom exception makes sense if it is intended to be used in the flow control.
In this case it is really abnormal to try to start the container which has not connection with its parent.
Therefore such a logic is wrong and has to be fixed in the target project.
So, `IllegalStateException` would just do whatever we need to indicated is wrong.

I'm not sure in your use-cases explanation, but looking into behavior one more time, I feel like we have to modify `ConcurrentMessageListenerContainer.doStop(final Runnable callback, boolean normal)` to call `destroy()` on all those child containers before `this.containers.clear();`.
That  `destroy()` should call not only `stop();` as it is right now in the `MessageListenerContainer`, but also mark some flag like `destroyed`.
Then when we try to call `start()` on that child container this new flag would give us the mentioned `IllegalStateException` indicating that container cannot be restarted after being destroyed.

Does that make sense to you?

Thank you for being patient with my review!

Reply:
Thank you. I have taken the core idea from your comments and modified the logic to solve this issue. I have planned to use `containerAllowedToStart `flag to verify if the container is allowed to start. Please consider. 

I have not used here `destroy` here for the following reasons.

1) `destroy` does not support to take the callback. 
2) A race could occur to occur to start the container immediately after stopped, if destroy is called before clearing the containers and after stopping the containers.